packages/slice-machine/src/features/onboarding/OnboardingProvider.tsx

```diff
@@ -93,3 +93,16 @@ export const useOnboardingContext = () => {
 
   return context;
 };
+
+export const useOnOnboardingComplete = (handler: () => void) => {
+  const { isComplete } = useOnboardingContext();
+  const previousIsCompleteRef = useRef<boolean>();
+
+  useEffect(() => {
+    previousIsCompleteRef.current = isComplete;
+  });
+
+  if (previousIsCompleteRef.current !== isComplete && isComplete) {
+    handler();
+  }
```

Comment:
we shouldn't read a ref in render, and we shouldn't perform a side effect in render

what are we trying to accomplish here? there's probably a more idiomatic way of doing it

Reply:
> we shouldn't read a ref in render

Why not?

This expression is to make sure if the handler is only called if the value of `isComplete` changes to `true`, and no subsequent calls are made if a re-render happens. I can use a `useEffect`, I just don't like how eslint always yells at you when I'm sure I don't want it to be triggered when the `handler` changes.

Reply:
I think I sent you the docs last week: https://react.dev/reference/react/useRef#caveats

>Do not write or read ref.current during rendering, except for [initialization.](https://react.dev/#avoiding-recreating-the-ref-contents) This makes your component’s behavior unpredictable.

You can do this in a use effect by passing a stable handler, but the idiomatic way is to execute the callback in an event handler. That's why I suggested passing it to the context in the other comment.

Reply:
Yeah I remember that link, but it was regarding accessing the current right after the definition. I don't think referencing refs in render time is a bad thing if you're sure you don't want it to change or trigger a re-render.